From c82886392315e33e0b14589bb435303122f34d23 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:09:02 -0400
Subject: [usb] cdc-acm: fix runtime PM for control messages

Message-id: <1410887463-20674-471-git-send-email-dzickus@redhat.com>
Patchwork-id: 95634
O-Subject: [RHEL7 PATCH 470/591] USB: cdc-acm: fix runtime PM for control messages
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit bae3f4c53585e9a170da9436e0f06919874bda9a
Author: Johan Hovold <jhovold@gmail.com>
Date:   Mon May 26 19:23:39 2014 +0200

    USB: cdc-acm: fix runtime PM for control messages

    Fix runtime PM handling of control messages by adding the required PM
    counter operations.

    Fixes: 11ea859d64b6 ("USB: additional power savings for cdc-acm devices
    that support remote wakeup")

    Cc: <stable@vger.kernel.org> # v2.6.27
    Signed-off-by: Johan Hovold <jhovold@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/class/cdc-acm.c b/drivers/usb/class/cdc-acm.c
index 5641925..2258827 100644
--- a/drivers/usb/class/cdc-acm.c
+++ b/drivers/usb/class/cdc-acm.c
@@ -122,13 +122,23 @@ static void acm_release_minor(struct acm *acm)
 static int acm_ctrl_msg(struct acm *acm, int request, int value,
        void *buf, int len)
 {
- int retval = usb_control_msg(acm->dev, usb_sndctrlpipe(acm->dev, 0),
+ int retval;
+
+ retval = usb_autopm_get_interface(acm->control);
+ if (retval)
+  return retval;
+
+ retval = usb_control_msg(acm->dev, usb_sndctrlpipe(acm->dev, 0),
   request, USB_RT_ACM, value,
   acm->control->altsetting[0].desc.bInterfaceNumber,
   buf, len, 5000);
+
  dev_dbg(&acm->control->dev,
    "%s - rq 0x%02x, val %#x, len %#x, result %d\n",
    __func__, request, value, len, retval);
+
+ usb_autopm_put_interface(acm->control);
+
  return retval < 0 ? retval : 0;
 }
 
-- 
1.7.1